src/audio/google/google_ctc_audio_processing.c

```diff
@@ -0,0 +1,357 @@
+		}
+	}
+	src = audio_stream_wrap(source, src + samples_to_process);
+	input_buffers->consumed += audio_stream_frame_bytes(source) * samples_to_process / n_ch;
+}
+
+static void ctc_s24_default(struct google_ctc_audio_processing_comp_data *cd,
+			    const struct audio_stream *source,
+			    struct audio_stream *sink,
+			    struct input_stream_buffer *input_buffers,
+			    struct output_stream_buffer *output_buffers,
+			    uint32_t frames)
+{
+	int n_ch = audio_stream_get_channels(source);
+	int samples = frames * n_ch;
+
+	int32_t *src = audio_stream_get_rptr(source);
+	int32_t *dest = audio_stream_get_wptr(sink);
+
+	int samples_to_process = MIN(samples, audio_stream_samples_without_wrap_s24(source, src));
+
+	for (int i = 0; i < samples_to_process; ++i) {
+		cd->input[cd->input_samples++] = convert_int32_to_float(src[i]);
+		if (cd->input_samples == cd->num_frames * n_ch) {
+			GoogleCtcAudioProcessingProcess(cd->state, cd->input, cd->output,
+							cd->num_frames, n_ch);
+
+			cd->input_samples = 0;
+			for (int i = 0; i < cd->num_frames * n_ch; ++i)
+				dest[i] = convert_float_to_int32(cd->output[i]);
+			dest = audio_stream_wrap(sink, dest + cd->num_frames * n_ch);
+			output_buffers->size += audio_stream_frame_bytes(sink) * cd->num_frames;
+		}
+	}
+	src = audio_stream_wrap(source, src + samples_to_process);
+	input_buffers->consumed += audio_stream_frame_bytes(source) * samples_to_process / n_ch;
+}
+
+static void ctc_s32_default(struct google_ctc_audio_processing_comp_data *cd,
+			    const struct audio_stream *source,
+			    struct audio_stream *sink,
+			    struct input_stream_buffer *input_buffers,
+			    struct output_stream_buffer *output_buffers,
+			    uint32_t frames)
+{
+	int n_ch = audio_stream_get_channels(source);
+	int samples = frames * n_ch;
+
+	int32_t *src = audio_stream_get_rptr(source);
+	int32_t *dest = audio_stream_get_wptr(sink);
+
+	int samples_to_process = MIN(samples, audio_stream_samples_without_wrap_s32(source, src));
+
+	for (int i = 0; i < samples_to_process; ++i) {
+		cd->input[cd->input_samples++] = convert_int32_to_float(src[i]);
+		if (cd->input_samples == cd->num_frames * n_ch) {
+			GoogleCtcAudioProcessingProcess(cd->state, cd->input, cd->output,
+							cd->num_frames, n_ch);
+
+			cd->input_samples = 0;
+			for (int i = 0; i < cd->num_frames * n_ch; ++i)
+				dest[i] = convert_float_to_int32(cd->output[i]);
+			dest = audio_stream_wrap(sink, dest + cd->num_frames * n_ch);
+			output_buffers->size += audio_stream_frame_bytes(sink) * cd->num_frames;
+		}
+	}
+	src = audio_stream_wrap(source, src + samples_to_process);
+	input_buffers->consumed += audio_stream_frame_bytes(source) * samples_to_process / n_ch;
+}
+
+static int ctc_free(struct processing_module *mod)
+{
+	struct google_ctc_audio_processing_comp_data *cd = module_get_private_data(mod);
+
+	if (cd) {
+		rfree(cd->input);
+		rfree(cd->output);
+		GoogleCtcAudioProcessingFree(cd->state);
```

Comment:
isn't this missing `rfree(cd)` here?

Reply:
also the new handler should be designed such that if we don't have cd then init failed and free shouldn't need to be called.

Reply:
Added `rfree(cd)`.